agent/engine/task_manager.go

```diff
@@ -159,31 +138,110 @@ func (mtask *managedTask) overseeTask() {
-	taskCredentialsID := mtask.GetCredentialsID()
-	if taskCredentialsID != "" {
-		mtask.engine.credentialsManager.RemoveCredentials(taskCredentialsID)
-	}
+	mtask.cleanupCredentials()
 	if mtask.StopSequenceNumber != 0 {
 		llog.Debug("Marking done for this sequence", "seqnum", mtask.StopSequenceNumber)
 		mtask.engine.taskStopGroup.Done(mtask.StopSequenceNumber)
 	}
 	mtask.cleanupTask(mtask.engine.cfg.TaskCleanupWaitDuration)
 }
 
+// emitCurrentStatus emits a container event for every container and a task
+// event for the task
 func (mtask *managedTask) emitCurrentStatus() {
 	for _, container := range mtask.Containers {
 		mtask.engine.emitContainerEvent(mtask.Task, container, "")
 	}
 	mtask.engine.emitTaskEvent(mtask.Task, "")
 }
 
+// waitForResources waits for host resources to become available to start the
+// task. This involves waiting for previous stops to complete so the resources
+// become free.
+func (mtask *managedTask) waitForResources() {
+	llog := log.New("task", mtask.Task)
+	if mtask.StartSequenceNumber != 0 && !mtask.GetDesiredStatus().Terminal() {
+		llog.Debug("Waiting for any previous stops to complete", "seqnum", mtask.StartSequenceNumber)
+		othersStopped := make(chan bool, 1)
+		go func() {
+			mtask.engine.taskStopGroup.Wait(mtask.StartSequenceNumber)
+			othersStopped <- true
+		}()
+		for !mtask.waitEvent(othersStopped) {
+			if mtask.GetDesiredStatus().Terminal() {
+				// If we end up here, that means we received a start then stop for this
+				// task before a task that was expected to stop before it could
+				// actually stop
+				break
+			}
+		}
+		llog.Debug("Wait over; ready to move towards status: " + mtask.GetDesiredStatus().String())
+	}
+	return
+}
+
+// waitSteady waits for a task to leave steady-state by waiting for a new
+// event, or a timeout.
+func (mtask *managedTask) waitSteady() {
+	llog := log.New("task", mtask.Task)
+	llog.Debug("Task at steady state", "state", mtask.GetKnownStatus().String())
+
+	maxWait := make(chan bool, 1)
+	timer := mtask.time().After(steadyStateTaskVerifyInterval)
+	go func() {
+		<-timer
+		maxWait <- true
+	}()
+	timedOut := mtask.waitEvent(maxWait)
+
+	if timedOut {
+		llog.Debug("Checking task to make sure it's still at steadystate")
+		go mtask.engine.CheckTaskState(mtask.Task)
+	}
+	return
+}
+
+// cleanupCredentials removes credentials for a stopped task
+func (mtask *managedTask) cleanupCredentials() {
+	taskCredentialsID := mtask.GetCredentialsID()
+	if taskCredentialsID != "" {
+		mtask.engine.credentialsManager.RemoveCredentials(taskCredentialsID)
+	}
+	return
+}
+
+// waitEvent waits for any event to occur. If an event occurs, the appropriate
+// handler is called. If the event is the passed in channel, it will return the
+// value written to the channel, otherwise it will return false.
+func (mtask *managedTask) waitEvent(stopWaiting <-chan bool) bool {
+	log.Debug("Waiting for event for task", "task", mtask.Task)
```

Comment:
I'd rather not. I replaced things with `seelog` when it was easy but didn't want to get too far down that rabbit hole in this specific PR.